osu.Framework/Bindables/BindableNumber.cs

```
@@ -2,33 +2,28 @@
 // See the LICENCE file in the repository root for full licence text.
 
 using System;
-using System.Collections.Generic;
 using System.Globalization;
 
 namespace osu.Framework.Bindables
 {
     public abstract class BindableNumber<T> : Bindable<T>, IBindableNumber<T>
-        where T : struct, IComparable, IConvertible
+        where T : struct, IComparable<T>, IConvertible, IEquatable<T>
     {
         static BindableNumber()
         {
-            // check supported types against provided type argument.
-            var allowedTypes = new HashSet<Type>
-            {
-                typeof(sbyte),
-                typeof(byte),
-                typeof(short),
-                typeof(ushort),
-                typeof(int),
-                typeof(uint),
-                typeof(long),
-                typeof(ulong),
-                typeof(float),
-                typeof(double)
-            };
-
-            if (!allowedTypes.Contains(typeof(T)))
-                throw new ArgumentException(
+            // Directly comparing typeof(T) to type literal is recognized pattern of JIT and very fast.
+            // Just a pointer comparison.
+            if (typeof(T) != typeof(sbyte) &&
+                typeof(T) != typeof(byte) &&
+                typeof(T) != typeof(short) &&
+                typeof(T) != typeof(ushort) &&
+                typeof(T) != typeof(int) &&
+                typeof(T) != typeof(uint) &&
+                typeof(T) != typeof(long) &&
+                typeof(T) != typeof(ulong) &&
+                typeof(T) != typeof(float) &&
+                typeof(T) != typeof(double))
```

Not sure this is worth it for one execution per potential generic type? I'd like to see benchmarks here too.

---

This runs per type instead of per instance, so there's no proper benchmark.
However, as [FxCop documentation](https://docs.microsoft.com/en-us/visualstudio/code-quality/ca1810?view=vs-2019#rule-description) says, static constructors do have performance impact on every method.
If the type can't be inherited outside, than the static check should only run on debug builds, like corefx does.

---

I mean I can imagine this type of check would be faster than the original `HashSet`, since it is allocated as a local every time the initializer is called. I'm not sure it *still* would be faster if the `HashSet` was moved out to a static field. I *imagine* it could (`GetHashCode()` will still probably be slower than the pointer check), but I don't *know* for sure, that's why I asked. I'm mostly griping on this since I don't necessarily like how the code *looks*, but I'm happy to be proved wrong.

Point definitely taken on the second part - if I understand correctly the type check _will_ only run once, but the usage of the initializer _itself_ comes with a performance penalty. In that case I wouldn't be completely opposed to guarding it with a preprocessor directive myself, but I'm not sure if it's a thing that's really done much around here, as it has its drawbacks, too.

---

Did a simple benchmark, and the result is 13.6376ns(HashSet) vs 0.6852ns(Direct compare) (~20x), even if I've picked the last one in the order.
For value types, I think JIT reduces the type check to constant.

---

The JIT will compile a separate generic method for every struct that's T. It will also evaluate all conditions like `if (typeof(T) == typeof(SomeType)` during compilation, not at runtime.
 After evaluating all the conditions that are JIT constant it will only take branches that are relevant for whatever the T is and then remove the checks from the compiled version (since they've been evaluated during compilation).
This means that testing for `typeof(T)` and `sizeof(T)` is free for structs and you can throw an exception when either the size or the type are incorrect without taking any performance hit (as the branch that throws an exception will be removed from the compiled method for correct sizes/types).

---

The JIT will compile a separate generic method for every struct that's T. It will also evaluate all conditions like `if (typeof(T) == typeof(SomeType)` during compilation, not at runtime.
 After evaluating all the conditions that are JIT constant it will only take branches that are relevant for whatever the T is and then remove the checks from the compiled version (since they've been evaluated during compilation).
This means that testing for `typeof(T)` and `sizeof(T)` is free for structs and you can throw an exception when either the size or the type are incorrect without taking any performance hit (as the branch that throws an exception will be removed from the compiled method for correct sizes/types).

---

This can be moved into the constructor and the static constructor can be removed.
This isn't even a pointer comparison, this is done by the JIT when compiling the generic.
If T is one of the allowed types the whole check will be removed and replaced by the appropriate branch (with the branching removed).
Otherwise it will only include the branch for types that are incorrect which means it will directly throw an exception.

---

@KoziLord you are right. [JIT eliminates the ctor into a nop](https://sharplab.io/#v2:C4LghgzgtgPgAgJgIwFgBQcAMACAygCzACcAHAGTACMA6AJQFcA7YASygFMBuddAbQCkWwAOLtG7IiwDGACmABPEuwD2AMxmV5wdgEodAXXRwAzNkTZhAHgAqAPnQBvdNhdnTwmTueunaV/+wWVWw5RRV1ax1sAEIAXmwFJTUNLV1sADJ07wCAxPCZSJj4vOSWZj1snNdgfCJlAHdsRnoAGxbuP1cAXx40LqA===).